packages/firestore/src/local/local_store.ts

```diff
@@ -708,6 +680,80 @@ class LocalStoreImpl implements LocalStore {
       });
   }
 
+  /**
+   * Populates document change buffer with documents from backend or a bundle.
+   * Returns the document changes resulting from applying those documents.
+   *
+   * @param txn Transaction to use to read existing documents from storage.
+   * @param documentBuffer Document buffer to collect the resulted changes to be
+   *        applied to storage.
+   * @param documents Documents to be applied.
+   * @param globalVersion A `SnapshotVersion` representing the read time if all
+   *        documents have the same read time.
+   * @param documentVersions A DocumentKey-to-SnapshotVersion map if documents
+   *        have their own read time.
+   *
+   * Note: this function will use `documentVersions` if it is defined;
+   * when it is not defined, resorts to `globalVersion`.
+   */
+  populateDocumentChangeBuffer(
+    txn: PersistenceTransaction,
+    documentBuffer: RemoteDocumentChangeBuffer,
+    documents: MaybeDocumentMap,
+    globalVersion: SnapshotVersion,
+    // TODO(wuandy): We could add `readTime` to MaybeDocument instead to remove
+    // this parameter.
+    documentVersions: DocumentVersionMap | undefined
+  ): PersistencePromise<MaybeDocumentMap> {
+    let updatedKeys = documentKeySet();
+    documents.forEach(k => (updatedKeys = updatedKeys.add(k)));
+    return documentBuffer.getEntries(txn, updatedKeys).next(existingDocs => {
+      let changedDocs = maybeDocumentMap();
+      documents.forEach((key, doc) => {
+        const existingDoc = existingDocs.get(key);
+        const docReadTime = documentVersions?.get(key) ?? globalVersion;
```

Comment:
Nit: I probably suggested this, but this is functionally equivalent to `documentVersions?.get(key) || globalVersion`, which generates a lot less code. `??` is only needed if you only want to check for undefined or null and explicitly want to not fallback on  "" (empty string), `0` or false.

```
if (doc?.version || 'foo')

=> 

if ((doc === null || doc === void 0 ? void 0 : doc.version) || 'foo')
 

if (doc?.version ?? 'foo')

=>

if ((_a = doc === null || doc === void 0 ? void 0 : doc.version) !== null && _a !== void 0 ? _a : 'foo')
```

As an example for where `??` works but `||` doesn't, consider:

```
class Settings {
   enableAmazingFeature?: boolean;
}

 // This is now always true
if (settings.enableAmazingFeature || true) {}
// This is true iff enableAmazingFeature=undefined or enableAmazingFeature=true
if (settings.enableAmazingFeature ?? true) {}
```

Reply:
I see, applied. Thanks!